OilPaintingProcessor memory usage improvements

Sergio0694 opened the pull request:
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [X] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [X] I have provided test coverage for my change (where applicable)

### Description
The `OilPaintingProcessor` was allocating 4 arrays for each single pixel being processed. I've switched to a single, shared `float[]` array rented from the used configuration, and I'm using that as temporary data for all the 4 processing bins. In doing so, I've also switched to `Unsafe.Add` for all the read/write operations to the 4 bins, for that extra micro-optimization breeze 🙌

Just noticed those 4 throwaway arrays while working on my other pull request and I couldn't resist 🤣

<!-- Thanks for contributing to ImageSharp! -->

Sergio0694 commented:
@JimBobSquarePants Don't merge this one just yet even if it passes the tests, I have another minor optimization done in another commit but I'll wait before pushing that one so that you'll have the CI free for your work 👍🏻
Will probably push it in 12 hours or so, so that it'd be night over there anyway.

Sergio0694 commented:
@JimBobSquarePants I've pushed my other commit now, though all the tests are failing in the CI for some reasons, I suspect it's just the CI itself that's acting up as before.
The tests for the `OilPaintingEffect` all pass successfully on my end.
Just an FYI, this PR isn't a priority since it's just a couple tweaks anyway 😊

cc. @antonfirsov in case you want to review the changes in this PR, as you're the optimization guru 😄
Also I was wondering whether it'd be better to add some more comments in the "weirdest" parts, for future reference? I mean eg. in parts like [this](https://github.com/SixLabors/ImageSharp/blob/692c35b166ca7a18680b5efa6cd863ee48e01e11/src/ImageSharp/Processing/Processors/Effects/OilPaintingProcessor%7BTPixel%7D.cs#L90). It's by no means no SSE/AVX magic, but still 🤔

JimBobSquarePants commented:
@Sergio0694 I don't think it's all CI issues. Maybe try running your tests locally in 32bit mode.

https://ci.appveyor.com/project/six-labors/imagesharp/builds/29921783

Sergio0694 commented:
@JimBobSquarePants Ok, this is weird - I've cloned `ImageSharp` from scratch into a new folder, initialized the submodules and everything, just opened `master` in Visual Studio and ran the `OilPaintTest` class, and I got this:

![image](https://user-images.githubusercontent.com/10199417/71897178-16577400-3156-11ea-9ff5-71bffb566c7c.png)

Same exact issue in the branch for this PR in my fork, I might be missing something but I don't think those errors are caused by changes I made in my branch, since `master` is failing for me as well with the same errors in the official repo.

What's more, the error seems to be with the checks in the `RowInterval` constructor that's invoked by the `ParallelHelper.IterateRows` method - it's throwing an `ArgumentOutOfRangeException` 🤔

brianpopow commented:
> @JimBobSquarePants Ok, this is weird - I've cloned `ImageSharp` from scratch into a new folder, initialized the submodules and everything, just opened `master` in Visual Studio and ran the `OilPaintTest` class, and I still got the same exact errors:
> 
> ![image](https://user-images.githubusercontent.com/10199417/71897178-16577400-3156-11ea-9ff5-71bffb566c7c.png)
> 
> Same issue in the branch for this PR in my for - I might be missing something but I don't think those errors are caused by changes I made in my branch, since `master` is failing for me as well with the same errors in the official repo.
> 
> What's more, the error seems to be with the checks in the `RowInterval` constructor that's invoked by the `ParallelHelper.IterateRows` method - it's throwing an `ArgumentOutOfRangeException` 🤔

Hi Sergio,

i have tried to reproduce your issue, but on the current master all tests are working fine. Also in the current master, i only see 4 oil paint tests:

![OilPaintTests](https://user-images.githubusercontent.com/38701097/71919891-f93a9980-3185-11ea-9324-bb940487ab04.jpg)

Sergio0694 commented:
@brianpopow Mmh... I had literally just cloned the entire repo into a new folder before running those tests, and I'm seeing the same results in my fork too. I have to say I'm very confused at this point 🤔
I might try this again on another PC entirely too at this point, just to be extra safe.

Sergio0694 commented:
@JimBobSquarePants I'm not really sure about what's going on here - I've merged the latest changes from `master` and tried to run the tests locally:

![image](https://user-images.githubusercontent.com/10199417/72078945-57888900-32fa-11ea-9428-4434b2826542.png)

Now I only see those 4 that @brianpopow mentioned, and they pass just fine for me here.
I might try to clone the repo again completely from scratch and just copy/paste the changes I made in the `OilPaintProcessor` there, but I really wonder what's happening in this PR 🤔

Sergio0694 commented:
So, I've merged all the changes from `master`, and while I was at it I also managed to throw in some more memory optimizations (in particular I've reduced the number of times the temporary buffer is rented from `numRows*columns` in each parallel task, to just `1`. 🚀

But, I can't seem to be able to figure out what's wrong with the tests. I've tried to clone `ImageSharp` from scratch, run the `OilPainTest` class, and again I got this.

![image](https://user-images.githubusercontent.com/10199417/73135359-6af95b00-4041-11ea-8f04-9badaec51a19.png)

The fact that `master` does build fine in the CI makes me think that somehow that could be caused by something wrong on my end, **but** then again this PR fails to build on the CI as well 🤷‍♂ 

I should add though that at least the "full image" tests run fine in this PR, it's just some of the "in box" ones that file. Just like with the ones in `master`, when I run them locally.

I have to say I'm a bit confused, I wouldn't mind someone taking a look here 😅

Sergio0694 commented:
@brianpopow Ah, I forgot they had already given the green light to C# 8 features in `ImageSharp`, otherwise I would've used the new using statements from the start 😄

Why did you revert https://github.com/SixLabors/ImageSharp/commit/692c35b166ca7a18680b5efa6cd863ee48e01e11 though? 🤔

brianpopow commented:
@Sergio0694 hi Sergio. I was able to reproduce the issue with your current branch. The error was introduced with the commit 692c35b166ca7a18680b5efa6cd863ee48e01e11 (I hope its ok that i reverted it). It seems that the memory was corrupt. This needs to be reviewed again, i could not spot the  exact error, but im sure this is how it was introduced.

Sergio0694 commented:
Hey @brianpopow - ah, sure, if that commit was the culprit then that's perfectly ok! 👍
I'll see if I can reintroduce that change though, since that was a pretty decent optimization. Also, I really wonder what's going on on my end at this point, since those tests are failing on my machine even on `master`, and that obviously can't be right.

Anyway, thank you so much for taking the time to look into this, I appreciate it! 😊

antonfirsov commented:
@Sergio0694 if I revert @brianpopow's revert commit, the memory corruption is even worse on my PC:
```
dotnet test --no-build -c Debug -f netcoreapp3.1 --filter FullyQualifiedName~OilPaintTest
```


<detail>

```
[xUnit.net 00:00:03.48]     InBox<Rgba32>(provider: Bmp/Car.bmp[Rgba32], levels: 15, brushSize: 10) [FAIL]
The active test run was aborted. Reason: Test host process crashed : Fatal error. Fatal error. Fatal error. Fatal error. Fatal error. Fatal error. Fatal error. Fatal error. Internal CLR error. (0x80131506)
Internal CLR error. (0x80131506)
Internal CLR error. (0x80131506)
Internal CLR error. (0x80131506)
Internal CLR error. (0x80131506)
   at System.Buffers.IMemoryOwner`1[[System.Single, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Memory()
   at System.IDisposable.Dispose()
   at System.Buffers.IMemoryOwner`1[[System.Single, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Memory()
Internal CLR error. (0x80131506)
   at SixLabors.ImageSharp.Processing.Processors.Effects.OilPaintingProcessor`1+<>c__DisplayClass2_1[[SixLabors.ImageSharp.PixelFormats.Rgba32, SixLabors.ImageSharp, Version=0.0.1.0, Culture=neutral, PublicKeyToken=null]].<OnFrameApply>b__0(SixLabors.ImageSharp.Memory.RowInterval, System.Memory`1<System.Numerics.Vector4>)
   at System.Buffers.IMemoryOwner`1[[System.Single, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Memory()
   at System.Buffers.IMemoryOwner`1[[System.Single, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Memory()
   at SixLabors.ImageSharp.Processing.Processors.Effects.OilPaintingProcessor`1+<>c__DisplayClass2_1[[SixLabors.ImageSharp.PixelFormats.Rgba32, SixLabors.ImageSharp, Version=0.0.1.0, Culture=neutral, PublicKeyToken=null]].<OnFrameApply>b__0(SixLabors.ImageSharp.Memory.RowInterval, System.Memory`1<System.Numerics.Vector4>)
   at SixLabors.ImageSharp.Processing.Processors.Effects.OilPaintingProcessor`1+<>c__DisplayClass2_1[[SixLabors.ImageSharp.PixelFormats.Rgba32, SixLabors.ImageSharp, Version=0.0.1.0, Culture=neutral, PublicKeyToken=null]].<OnFrameApply>b__0(SixLabors.ImageSharp.Memory.RowInterval, System.Memory`1<System.Numerics.Vector4>)
   at SixLabors.ImageSharp.Advanced.ParallelUtils.ParallelHelper+<>c__DisplayClass2_0`1[[System.Numerics.Vector4, System.Numerics.Vectors, Version=4.1.6.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].<IterateRowsWithTempBuffer>b__0(Int32)
Internal CLR error. (0x80131506)
   at System.Buffers.IMemoryOwner`1[[System.Single, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Memory()
   at SixLabors.ImageSharp.Advanced.ParallelUtils.ParallelHelper+<>c__DisplayClass2_0`1[[System.Numerics.Vector4, System.Numerics.Vectors, Version=4.1.6.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].<IterateRowsWithTempBuffer>b__0(Int32)
   at SixLabors.ImageSharp.Processing.Processors.Effects.OilPaintingProcessor`1+<>c__DisplayClass2_1[[SixLabors.ImageSharp.PixelFormats.Rgba32, SixLabors.ImageSharp, Version=0.0.1.0, Culture=neutral, PublicKeyToken=null]].<OnFrameApply>b__0(SixLabors.ImageSharp.Memory.RowInterval, System.Memory`1<System.Numerics.Vector4>)
   at SixLabors.ImageSharp.Processing.Processors.Effects.OilPaintingProcessor`1+<>c__DisplayClass2_1[[SixLabors.ImageSharp.PixelFormats.Rgba32, SixLabors.ImageSharp, Version=0.0.1.0, Culture=neutral, PublicKeyToken=null]].<OnFrameApply>b__0(SixLabors.ImageSharp.Memory.RowInterval, System.Memory`1<System.Numerics.Vector4>)
   at SixLabors.ImageSharp.Advanced.ParallelUtils.ParallelHelper+<>c__DisplayClass2_0`1[[System.Numerics.Vector4, System.Numerics.Vectors, Version=4.1.6.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].<IterateRowsWithTempBuffer>b__0(Int32)
   at SixLabors.ImageSharp.Advanced.ParallelUtils.ParallelHelper+<>c__DisplayClass2_0`1[[System.Numerics.Vector4, System.Numerics.Vectors, Version=4.1.6.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].<IterateRowsWithTempBuffer>b__0(Int32)
   at System.Threading.Tasks.Parallel+<>c__DisplayClass19_0`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForWorker>b__1(System.Threading.Tasks.RangeWorker ByRef, Int32, Boolean ByRef)
   at System.Threading.Tasks.Parallel+<>c__DisplayClass19_0`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForWorker>b__1(System.Threading.Tasks.RangeWorker ByRef, Int32, Boolean ByRef)   at System.Threading.Tasks.TaskReplicator+Replica`1[[System.Threading.Tasks.RangeWorker, System.Threading.Tasks.Parallel, Version=4.0.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].ExecuteAction(Boolean ByRef)   at System.Threading.Tasks.Parallel+<>c__DisplayClass19_0`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForWorker>b__1(System.Threading.Tasks.RangeWorker ByRef, Int32, Boolean ByRef)
   at SixLabors.ImageSharp.Advanced.ParallelUtils.ParallelHelper+<>c__DisplayClass2_0`1[[System.Numerics.Vector4, System.Numerics.Vectors, Version=4.1.6.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].<IterateRowsWithTempBuffer>b__0(Int32)   at SixLabors.ImageSharp.Advanced.ParallelUtils.ParallelHelper+<>c__DisplayClass2_0`1[[System.Numerics.Vector4, System.Numerics.Vectors, Version=4.1.6.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].<IterateRowsWithTempBuffer>b__0(Int32)   at System.Buffers.IMemoryOwner`1[[System.Single, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Memory()
   at System.Threading.Tasks.Parallel+<>c__DisplayClass19_0`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForWorker>b__1(System.Threading.Tasks.RangeWorker ByRef, Int32, Boolean ByRef)   at System.Threading.Tasks.TaskReplicator+Replica`1[[System.Threading.Tasks.RangeWorker, System.Threading.Tasks.Parallel, Version=4.0.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].ExecuteAction(Boolean ByRef)
Internal CLR error. (0x80131506)
```
</detail>

I wonder how should we proceed. Maybe merge it as-is for now, and try adding furter fixes later?

Sergio0694 commented:
@antonfirsov Yup, that's what I was thinking about, I'll give that a try!
It's weird though that there's quite a lot of unsafe indexing already, but all the tests are passing just fine right now. I wonder what's going on there. I have a few hours of free time, I'll let you know as soon as I have an update on this! 👍

Sergio0694 commented:
@antonfirsov Managed to fix all the issues and also added some more minor optimizations, I'm pretty happy about the implementation right now. I've added you as a reviewer, let me know what you think! 😊

Sergio0694 commented:
@antonfirsov I'll be honest, I really have no clue about the details of the underlying algorithms either, I've just focused on doing some general memory optimizations, trusting the tests as well 😅

antonfirsov commented:
@Sergio0694 been there so many times with jpeg!

JimBobSquarePants commented on src/ImageSharp/Processing/Processors/Effects/OilPaintingProcessor{TPixel}.cs:
```diff
@@ -52,82 +54,107 @@ protected override void OnFrameApply(ImageFrame<TPixel> source)
+                (rows) =>
+                {
+                    // Allocate the reusable source row buffer, to enable vectorized bulk conversions
+                    using IMemoryOwner<Vector4> sourceRowBuffer = configuration.MemoryAllocator.Allocate<Vector4>(rowWidth);
```
I'm not a big fan of the new `using` declaration as I believe it reduces readability in many situations and can promote maintaining object lifespans for longer than they should be. 

If they were to be used (haha) then I would restrict the environment to small methods with a single instance. That is not possible however so I'm very tempted to add a rule that disallows them.

antonfirsov replied:
So let the fight begin! Me and @brianpopow like them 😄

Sergio0694 replied:
I have to say I like them very much too, though I can see how in some cases they might make the code less intuitive at first. For now I've removed them in https://github.com/SixLabors/ImageSharp/pull/1068/commits/c89e0b1e5a1ff9e266cb3a8ef5febaabbde13059, and if we change our mind on this we can always refactor this later on, it's just a matter of removing two brackets anyway 👍

JimBobSquarePants commented on src/ImageSharp/Processing/Processors/Effects/OilPaintingProcessor{TPixel}.cs:
```diff
@@ -52,82 +54,107 @@ protected override void OnFrameApply(ImageFrame<TPixel> source)
-                                    int maxIntensity = 0;
-                                    int maxIndex = 0;
+            var workingRect = Rectangle.FromLTRB(startX, startY, endX, endY);
+            ParallelHelper.IterateRows(
```
`IterateRowsWithTempBuffer`

Sergio0694 replied:
I'm not using that on purpose, since I need two buffers anyway. Plus, the buffer I need needs to be as wide as the entire row, not just the target rectangle area, so using that helper method is not possible in this case.
Should I add a detailing comment there to explain the rationale?

JimBobSquarePants replied:
Ah ok, sorry. I didn't realize it was only the size of the working rectangle.

Sergio0694 replied:
No worries, that was actually a very good point and I had the same doubt at first. You were absolutely right to ask me to clarify that, and I think having a more detailed comment in there will definitely come in handy when (if) someone else will look back at this code again in a few months 👍

Sergio0694 commented:
Yeah! The RC1 is one tiny step closer! 🚀🍻
